src/Equinox/Decider.fs

```diff
@@ -290,3 +288,34 @@ and internal SyncContext<'state> =
             member _.Version = token.version
             member _.StreamEventBytes = match token.streamBytes with -1L -> ValueNone | b -> ValueSome b
             member _.CreateMemento() = token, state }
+
+[<Struct>]
+type Target =
+    { categoryName : string; streamId : string }
+    static member Render(categoryName, streamId) = String.Concat(categoryName, '-', streamId)
+    override x.ToString() = Target.Render(x.categoryName, x.streamId)
+
+module Target =
+    module Internal =
+        /// Throws if a candidate category includes a '-', is null, or is empty
+        let inline validateCategory (rawCategory : string) =
+            if rawCategory |> String.IsNullOrEmpty then invalidArg "rawCategory" "may not be null or empty"
+            if rawCategory.IndexOf '-' <> -1 then invalidArg "rawCategory" "may not contain embedded '-' symbols"
+        /// Throws if a candidate id element includes a '_', is null, or is empty
+        let inline validateElement (rawElement : string) =
+            if rawElement |> String.IsNullOrEmpty then invalidArg "rawElement" "may not contain null or empty components"
+            if rawElement.IndexOf '_' <> -1 then invalidArg "rawElement" "may not contain embedded '_' symbols"
+        /// Low level helper used to gate ingestion from a canonical form
+        /// Does NOT validate the {streamId} portion wrt numbers of embedded `_` or `-` chars
+        let create categoryName streamId =
+            validateCategory categoryName
+            { categoryName = categoryName; streamId = streamId }
+        let toString (x : Target) = string x
+
+    let private streamIdOfElement f id = let t = f id in Internal.validateElement t; t
+    let gen categoryName f id = Internal.create categoryName (streamIdOfElement f id)
+    let private streamIdOfElements (elements : string seq) : string =
+        for x in elements do Internal.validateElement x
+        String.Join("_", elements)
+    let gen2 categoryName f f2 struct (id1, id2) = Internal.create categoryName (streamIdOfElements (seq { yield f id1; yield f2 id2 }))
+    let gen3 categoryName f f2 f3 struct (id1, id2, id3) = Internal.create categoryName (streamIdOfElements (seq { yield f id1; yield f2 id2; yield f3 id3 }))
```

Comment:
I think it'd be great to have a

```fsharp
let withContext target resolve ctx args = target args |> resolve ctx
```

Reply:
Added it, not sure I get it; Maybe if we have a test that could illustrate it, I could see if some other changes could render it unnecessary

Reply:
The two liner test would be:

```fsharp
let target = Target.gen "Category" guidN |> Target.withContext
let create resolve = Service(target resolve)
```

Not sure if it's actually beneficial over

```fsharp
let target = Target.gen "Category" guidN |> Target.withContext
let create resolve = Service(fun ctx id -> resolve ctx (target id))
```

Reply:
(Resolving this, lets continue in the open one)